### PR TITLE
let port to send retry packet to validate client

### PR DESCRIPTION
**Pull request by Sashan**
This is the draft PR just for reference to monitor progress in sprint. there is no intention to merge it.

My initial implementation did send retry packets from channel. it was too late, we need to send retry packet before channel gets created. I believe this is what needs to happen:
  - determine we need to validate client (TODO: introduce counter to monitor half-opened connections)
  - if we need to validate client we need to generate new connection id
  - we need to generate token (TODO: token needs protection against spoofing)
  - port sends retry packet, packet comes from source connection ID, client will use this source connection ID as destination connection ID for next initial packet
  - server then receives next initial packet with validation token
  - server checks if token is valid
  - if token is valid the server must create channel and bind it to source connection id. this is the detail where devil hides

existing function `port_on_new_conn()` creates channel and connection id one step. connection is bound to channel. here. also keep in mind connection id is object which gets inserted to look up table.

When dealing with address validation we don't want to create any persistent objects (channel and connection id). The draft PR here introduces function `port_bind_channel()`. It allows us to generate random connection ID at the time server sends a retry packet. Once we receive token back with next initial packet we create channel, and connection id object and insert it to look up table. (TODO: The insertion may fail, in this case we need to send RESET packet back, so client can quickly retry to connect).

The prototype still needs some love it looks like I'e messed up insertion of connection id to lookup table. I'm not able to find the channel I've created.

### CODE REVIEW

**Comment by Sashan** [comment]
to test the code use:
```
cd demos/quic/server
gdb --args ./server 4433 ../../../test/certs/servercert.pem ../../../test/certs/serverkey.pem
```

run server, add brekapoints to `port_bind_channel` where we create channel for validated client. continue the next packet from client should find channel using `ossl_quic_lcidm_lookup()` but it does not happen.

to trigger the breakpoint you need to poke to server using client:
```
test/quic_client_test
```
the goal here is to get SSL handshake working.

**Comment by mattcaswell** [comment]
> cd demos/quic/server

Umm...no such dir?

File: ssl/quic/quic_lcidm.c [hunk 1]
```diff
@@ -554,3 +600,17 @@ int ossl_quic_lcidm_debug_add(QUIC_LCIDM *lcidm, void *opaque,
     lcid_obj->type      = LCID_TYPE_NCID;
     return 1;
 }
+
+int ossl_quic_lcidm_get_unused_cid(QUIC_LCIDM *lcidm, QUIC_CONN_ID *cid)
+{
+    int i;
+
+    for (i = 0; i < 10; i++) {
+        if (lcidm_generate_cid(lcidm, cid))
```
Comment (mattcaswell) [inline ssl/quic/quic_lcidm.c:609]: All this does is generate a random cid. It doesn't add it to the lcidm. Don't you need to call `lcidm_conn_new_lcid` to actually add it? Compare with `lcidm_generate`

Comment (Sashan) [reply ssl/quic/quic_lcidm.c:609]: my intention is:
   - generate random lcid which is not used (happens in snippet above)
   - use the random lcid as source con id in retry packet
   - the next initial packet which carries validation token uses lcid as destination connection id
   - the `port_bind_channel()` function creates channel and inserts `lcid` (`dstConnId` from packet header) to `lcidm`
 
the bug must be hiding in the codepath which implements the last step above.

**Comment by Sashan** [comment]
update: the channel is actually found. the server accepts connection. then proceeds to call to `write()`. The server gets stuck in write. According to wireshark the server does not continue with handshake. Client keeps retrying with initial packet, but server does send its HANDSHAKE packet.

it looks like the channel is not able to perform a handshake.

**Comment by Sashan** [comment]
</snip>
> We can see that the next Initial packet shows a retry token that is 21 bytes long (the openssltoken string plus the trailing garbage, and the next packet in the datagram (which should be the TLS client hello) is not decoded properly

the trailing garbage is _original_ destination connection id where client did send its connection request.
see `port_send_retry()` function for details where validation token is constructed:
```
680
681     /* TODO: generate proper validation token */g
682     memcpy(token.token_buf, "openssltoken", sizeof("openssltoken") - 1);g
683g
684     token.token_odcid = client_hdr->dst_conn_id;g
685     token_len += token.token_odcid.id_len;g
686     integrity_tag = (char *)&token.token_odcid +g
687         token.token_odcid.id_len + sizeof(token.token_odcid.id_len);g
688     /*g
689      * 17.2.5.1 Sending a Retry packetg
690      *   dst ConnId is src ConnId we got from clientg
691      *   src ConnId comes from local conn ID managerg
692      */g
693     memset(&hdr, 0, sizeof(QUIC_PKT_HDR));g
694     hdr.dst_conn_id = client_hdr->src_conn_id;g
695     /*g
696      * this is the random connection ID, we expect client isg
697      * going to send the ID with next INITIAL packet whichg
698      * will also come with token we generate here.g
699      */g
700     ok = ossl_quic_lcidm_get_unused_cid(port->lcidm, &hdr.src_conn_id);g
701     if (ok == 0)g
702         return;g
703g
704     hdr.dst_conn_id = client_hdr->src_conn_id;g
705     hdr.type = QUIC_PKT_TYPE_RETRY;g
706     hdr.fixed = 1;g
707     hdr.version = 1;g
708     hdr.len = token_len;g
709     hdr.data = (unsigned char *)&token;g
710     ok = ossl_quic_calculate_retry_integrity_tag(port->engine->libctx,g
711                                                  port->engine->propq, &hdr,g
712                                                  &client_hdr->dst_conn_id,g
713                                                  integrity_tag);g
714     if (ok == 0)g
715         return;g
716g
717     hdr.token = (unsigned char *)&token;g
718     hdr.token_len = token_len;g
719g
720     msg[0].data = buffer;g
721     msg[0].peer = peer;g
722     msg[0].local = NULL;g
723     msg[0].flags = 0;g
724g
725     ok = WPACKET_init_static_len(&wpkt, buffer, sizeof(buffer), 0);g
726     if (ok == 0)g
727         return;g
728g
729     ok = ossl_quic_wire_encode_pkt_hdr(&wpkt, client_hdr->dst_conn_id.id_len,g
730                                        &hdr, NULL);g
731     if (ok == 0)g
732         return;g
733g
734     ok = WPACKET_get_total_written(&wpkt, &msg[0].data_len);g
735     if (ok == 0)g
736         return;g
737g
738     ok = WPACKET_finish(&wpkt);g
739     if (ok == 0)g
740         return;g
741g
742     BIO_sendmmsg(port->net_wbio, msg, sizeof(BIO_MSG), 1, 0, &written);g
```

the token is chain of 'openssltoken' _secret_ followed by original destination connection id.

**Comment by nhorman** [comment]
Ok, thank you @Sashan .  So that all makes sense then, and it seems that the problem is at least in part that the crypto packet in the retired datagram isn't getting decrypted properly.  Wireshark is indicating that the stream offset field is duplicated in the retried datagram, which may either be an error in Wireshark, or part of the cause to this issue. Will keep looking

**Comment by nhorman** [comment]
Actually I should word that differently.  Initial packets aren't encrypted, but we're trying.ti decrypt it, so something about the first non retried frames is putting the quic.port in a state where it thinks it's not the initial tls.client hello frame

**Comment by Sashan** [comment]
> Actually I should word that differently. Initial packets aren't encrypted, but we're trying.ti decrypt it, so something about the first non retried frames is putting the quic.port in a state where it thinks it's not the initial tls.client hello frame

I assume you mean initial packets which come with validation token (come as a _reply_ to retry packet server sent). The first such packet arriving at server creates a channel The packet is then dispatched to channel where it should be handled (it starts in `ch_rx()` function). So this is the place where I would start looking at, but things are more complicated because everything is running on behalf of `tick()`. Port just puts the packet to the queue which is consomued by channel (`ch_rx()`) function. The `ch_rx()` gets invoked from tick.

**Comment by nhorman** [comment]
yes, that is what I mean.  Currently, what (I think) is happening is that:
1) client sends an initial frame without a retry token
2) server sends a retry frame back
3) client sends the initial frame again
4) server accepts the frame in (3), and creates a channel in port_bind_channel, doing an hkdf deriviation for the el_set keys used to decode the crypto packet in the frame in (3)
5) server fails crypto packet decryption

I'm guessing that, in during channel creation (likely in ch_on_new_conn_common), when we call ossl_quic_provide_initial_secret, we do the hkdf derivation using the initial dcid rather than the scid provided in the retry packet, leading to decryption errors of the crypto packet, but I need to look more closely

**Comment by Sashan** [comment]
</snip>
> I'm guessing that, in during channel creation (likely in ch_on_new_conn_common), when we call ossl_quic_provide_initial_secret, we do the hkdf derivation using the initial dcid rather than the scid provided in the retry packet, leading to decryption errors of the crypto packet, but I need to look more closely
yes this was also my guess. I have not entirely eliminated that option yet. I've just pushed changes which print connection ids to stderr. I'm sorry I have not pushed those on Saturday. I hope they will be still useful to you.

**Comment by nhorman** [comment]
thank you!  I think we're on the right track here.  If I use this patch:
```
commit 0556ab493c1c75e3fe5f9b65d33cc9b56da3c147
Author: Neil Horman <nhorman@openssl.org>
Date:   Tue Nov 5 06:43:28 2024 -0500

    Generate initial secrets based on new dcid rather than odcid

diff --git a/ssl/quic/quic_channel.c b/ssl/quic/quic_channel.c
index b8b0b0743c..0e7b2db65f 100644
--- a/ssl/quic/quic_channel.c
+++ b/ssl/quic/quic_channel.c
@@ -3364,7 +3364,8 @@ static void ch_on_idle_timeout(QUIC_CHANNEL *ch)
 
 static int ch_on_new_conn_common(QUIC_CHANNEL *ch, const BIO_ADDR *peer,
                                  const QUIC_CONN_ID *peer_scid,
-                                 const QUIC_CONN_ID *peer_dcid)
+                                 const QUIC_CONN_ID *peer_dcid,
+                                 const QUIC_CONN_ID *peer_odcid)
 {
     /* Note our newly learnt peer address and CIDs. */
     ch->cur_peer_addr   = *peer;
@@ -3395,7 +3396,7 @@ static int ch_on_new_conn_common(QUIC_CHANNEL *ch, const BIO_ADDR *peer,
         return 0;
 
     /* Register the peer ODCID in the LCIDM. */
-    if (!ossl_quic_lcidm_enrol_odcid(ch->lcidm, ch, &ch->init_dcid))
+    if (!ossl_quic_lcidm_enrol_odcid(ch->lcidm, ch, peer_odcid == NULL ? peer_dcid : peer_odcid))
         return 0;
 
     /* Change state. */
@@ -3416,7 +3417,7 @@ int ossl_quic_channel_on_new_conn(QUIC_CHANNEL *ch, const BIO_ADDR *peer,
     if (!ossl_quic_lcidm_generate_initial(ch->lcidm, ch, &ch->cur_local_cid))
         return 0;
 
-    return ch_on_new_conn_common(ch, peer, peer_scid, peer_dcid);
+    return ch_on_new_conn_common(ch, peer, peer_scid, peer_dcid, NULL);
 }
 
 int ossl_quic_bind_channel(QUIC_CHANNEL *ch, const BIO_ADDR *peer,
@@ -3438,7 +3439,7 @@ int ossl_quic_bind_channel(QUIC_CHANNEL *ch, const BIO_ADDR *peer,
      * peer_odcid <=> is initial dst conn id chosen by peer in its
      * first initial packet we received without token.
      */
-    return ch_on_new_conn_common(ch, peer, peer_scid, peer_odcid);
+    return ch_on_new_conn_common(ch, peer, peer_scid, peer_dcid, peer_odcid);
 }
 
 SSL *ossl_quic_channel_get0_ssl(QUIC_CHANNEL *ch)
```
then the crypto packet in the init frame sent from the client in response to the retry frame gets decoded on the server correctly, as the initial secrets are generated based on the new, rather then the original dcid, which I believe was what the client was encrypting it with.  With this, the handshake continues, but runs into an error about a dcid mismatch, which I need to track down.  Also need to read the RFC to see if there is a proscribed behavior regarding what they initial keys should use as a secret value (RFC appears vague on that).

**Comment by nhorman** [comment]
ok, according to:
https://datatracker.ietf.org/doc/html/rfc9000#name-handling-a-retry-packet
the RFC says that re-sending the initial frame in response to a retry mandates that, because the dcid is changed to the id specified in the retry packet mandates that the initial namespace keys in the el_set be updated to reflect that, so it seems that the client side is doing the right thing, and the patch above, while a bit of a hack, is doing the right thing, so I think the next step is to figure out why we're getting a dcid mismatch error in the handshake and correct that

**Comment by nhorman** [comment]
Screenshot from 2024-11-05 08-49-13

Appears that the problem is that the client in the re-sent inital frame includes in the TLS handshake the initial source connection id transport parameter, but neglects to fill it out

**Comment by nhorman** [comment]
hmm, with that patch above we continue the handshake, but then fail because we don't encode the proper destination cid in the server response, after which the client closes the connection with a TRANSPORT_PARAM_ERROR.

This patch solves that:
```
commit ecfc56cfa1541380826ec4014db43c8dcfd22254
Author: Neil Horman <nhorman@openssl.org>
Date:   Tue Nov 5 10:52:52 2024 -0500

    ensure that ocid is sent after a retry frame causes new inital packet

diff --git a/ssl/quic/quic_channel.c b/ssl/quic/quic_channel.c
index 0e7b2db65f..4224996d72 100644
--- a/ssl/quic/quic_channel.c
+++ b/ssl/quic/quic_channel.c
@@ -1716,6 +1716,7 @@ static int ch_generate_transport_params(QUIC_CHANNEL *ch)
     WPACKET wpkt;
     int wpkt_valid = 0;
     size_t buf_len = 0;
+    int rc;
 
     if (ch->local_transport_params != NULL || ch->got_local_transport_params)
         goto err;
@@ -1733,8 +1734,14 @@ static int ch_generate_transport_params(QUIC_CHANNEL *ch)
         goto err;
 
     if (ch->is_server) {
-        if (!ossl_quic_wire_encode_transport_param_cid(&wpkt, QUIC_TPARAM_ORIG_DCID,
-                                                       &ch->init_dcid))
+        if (ch->odcid.id_len == 0)
+            rc = ossl_quic_wire_encode_transport_param_cid(&wpkt, QUIC_TPARAM_ORIG_DCID,
+                                                           &ch->init_dcid);
+        else
+            rc = ossl_quic_wire_encode_transport_param_cid(&wpkt, QUIC_TPARAM_ORIG_DCID,
+                                                           &ch->odcid);
+
+        if (rc == 0)
             goto err;
 
         if (!ossl_quic_wire_encode_transport_param_cid(&wpkt, QUIC_TPARAM_INITIAL_SCID,
@@ -3371,6 +3378,9 @@ static int ch_on_new_conn_common(QUIC_CHANNEL *ch, const BIO_ADDR *peer,
     ch->cur_peer_addr   = *peer;
     ch->init_dcid       = *peer_dcid;
     ch->cur_remote_dcid = *peer_scid;
+    ch->odcid.id_len = 0;
+    if (peer_odcid != NULL)
+        ch->odcid = *peer_odcid;
 
     /* Inform QTX of peer address. */
     if (!ossl_quic_tx_packetiser_set_peer(ch->txp, &ch->cur_peer_addr))
diff --git a/ssl/quic/quic_channel_local.h b/ssl/quic/quic_channel_local.h
index 7cafa109c4..232201c23e 100644
--- a/ssl/quic/quic_channel_local.h
+++ b/ssl/quic/quic_channel_local.h
@@ -106,7 +106,7 @@ struct quic_channel_st {
      * Randomly generated and required by RFC to be at least 8 bytes.
      */
     QUIC_CONN_ID                    init_dcid;
-
+    QUIC_CONN_ID                    odcid;
     /*
      * Client: The SCID found in the first Initial packet from the server.
      * Not valid for servers.
```
I'll push these to a scratch branch in a bit.

That gets the negotiation a little farther, but we now fail on another parameter error (which I'm guessing we never tested as we don't normally do retry packets).  With the above patch we encounter another parameter error, specifically a missing QUIC_TPARAM_RETRY_SCID param, which we never encode at all.  QUIC_TPARAM_RETRY_SCID is an extended param that is part of the QUIC retry offload draft RFC:
https://datatracker.ietf.org/doc/html/draft-ietf-quic-retry-offload-00

Apparently when this was coded we check for its presence on receiption, but we never do the encoding at all.  

Given that the above draft RFC expired in November of 2022, and was never updated or replaced, I'm inclined to just rip that code out, as it appears its not likely to get codified as a standard at all

**Comment by nhorman** [comment]
Based on discussion with @vavroch2010 I'm going to take the issues i've found and split them into separate issues so we can work on them in parallel.  I'll add the task list to the top of the issue associated with this PR

**Comment by Sashan** [comment]
You should also try client with cloadflare quic server, this is not build by default. If I remember correct the cloudflare server validates client. This is what you need to do to get build it and test it to capture quic packets
```
cd $YOUR_OPENSSL_ROOT
git clone --recursive https://github.com/cloudflare/quiche cloudflare-quiche
./Configure -d enable-external-tests enable-unstable-qlog
make 
make TESTS='test_external_cf_quiche' test
```

Once the above is done you can run it in your own which is useful if you need to watch client in debugger
```
cd $YOUR_OPENSSL_ROOT
export SRCTOP=`pwd`
export BLDTOP=`pwd`
export QUICHE_TARGET_PATH="$BLDTOP/quiche"
"$QUICHE_TARGET_PATH/debug/quiche-server" --cert "$SRCTOP/test/certs/servercert.pem" \
     --key "$SRCTOP/test/certs/serverkey.pem" --disable-gso \
     --http-version HTTP/0.9 --root "$SRCTOP" --no-grease --disable-hystart
```

The client is the one we use (`test/quic_client_test`)

**Comment by nhorman** [comment]
hey hey!  With my hacked up code from the 3 tasks listed [here](https://github.com/openssl/project/issues/715#issue-2362584446) quiche seems to work nicely!

**Comment by Sashan** [comment]
this is no longer relevant, the change will be delivered by #25890

**Pull request closed without merging**